magma/conversions.py

```diff
@@ -260,3 +260,112 @@ def replace(value, others: dict):
         return array(l)
     else:
         raise ValueError("replace can only be used with an Array, a Tuple, or Product")
+
+
+def _dispatch(cls, dispatch, *args, **kwargs):
+    mro = cls.mro()
+    for dispatch_cls in mro:
+        fn = dispatch.registry.get(dispatch_cls, None)
+        if fn is not None:
+            return fn(*args, **kwargs)
+    raise NotImplementedError()
+
+
+@functools.singledispatch
+def _as_bits(value):
+    """
+    Convert value to "flat" representation as an instance of m.Bits
+
+    For example, flatten a multi-dimensional array `m.Array[4,
+    m.Bits[5]]` into `m.Bits[20]`.
+
+    This is the inverse of `from_bits`
+    """
+    raise NotImplementedError()
+
+
+@_as_bits.register
+def _(value: Digital):
+    result = m.Bits[1]()
+    result[0] @= value
+    return result
+
+
+@_as_bits.register
+def _(value: Array):
+    if isinstance(value.T, Digital):
+        return Bits[len(value)](value.ts)
+    return functools.reduce(lambda x, y: x.concat(y),
+                            map(lambda x: as_bits(x), value.ts))
+
+
+@_as_bits.register
+def _(value: Bits):
+    return value
+
+
+@_as_bits.register
+def _(value: Tuple):
+    return functools.reduce(lambda x, y: x.concat(y),
+                            map(lambda x: as_bits(x), value.values()))
+
+
+def as_bits(value):
+    return _dispatch(type(value), _as_bits, value)
+
+
+@functools.singledispatch
+def _from_bits(cls, value):
+    """
+    Create value from a "flat" representation as an instance of m.Bits.
+
+    For example, pack an `m.Bits[20]` into an `m.Array[4, m.Bits[5]]`.
+
+    This is the inverse of `as_bits`
+    """
+    raise NotImplementedError()
+
+
+@_from_bits.register(Digital)
+def _(cls, value):
+    if not isinstance(value, Bits) and not len(value) == 1:
+        raise TypeError("Can only convert from Bits[1] to Bit")
+    return value[0]
+
+
+@_from_bits.register(Bits)
+def _(cls, value):
+    return value
+
+
+@_from_bits.register(Array)
+def _(cls, value):
+    if issubclass(cls.T, Digital):
+        if not len(cls) == len(value):
+            raise TypeError("Width mismatch")
+        return cls(value.ts)
+    child_length = cls.T.flat_length()
+    children = [
+        from_bits(cls.T,
+            value[i * child_length:(i + 1) * child_length]
+        )
+        for i in range(child_length)
+    ]
+    return cls(children)
+
+
+@_from_bits.register(Tuple)
+def _(cls, value):
+    children = []
+    offset = 0
+    for child in cls.fields:
+        child_length = child.flat_length()
+        children.append(
+            from_bits(child, value[offset:offset + child_length])
+        )
+        offset += child_length
+    return cls(*children)
+
+
+def from_bits(cls, value):
+    return _dispatch(cls, _from_bits, cls, value)
```

Comment:
I think we should expose (and document) `_from_bits` and `_as_bits` so users can define their own types that implement this interface (that was the benefit of having it in the type definition themselves).  While moving them into a separate conversions module for organization makes sense, I think we should still allow users to define custom implementations of this logic.

Reply:
I agree. I'm actually wondering why we shouldn't just use `flatten` and something like `reshape`? Or even better a generic `reinterpret` function?

IMO, right now there's many ways to do the same thing of converting one type to another in magma. We should just decide on one way and make that the standard. For now I'll leave this as is, but I agree overriding this behavior should be possible.